### DISCUSSION TITLE
Stopping Criteria for CNMF-E in Caiman

### DISCUSSION THREAD

**Discussion by bashsyr** [flatironinstitute/CaImAn#1241] [General]
Hello,
I'm writing my thesis on optimizing the parameters of Caiman and I'm  looking for an ansewr for this question for a while now but I didn't find any.
How does the CNMF-E algorithm stop interating and decides it's done (in Caiman and also in general)
I read that the algorithm would converge. But how does it converge and on which base?
I'd be gratefull for any further help
Thank you

**Comment by EricThomson** [comment]
Hi thanks for the great question! This is something I've thought about adding visualizations for in a notebook (e.g., we don't really discuss `maxIters` for the hierarchical alternating least squares algorithm). Unfortunately I don't have anything useful to say about it yet: this is pretty deep into the to-do list type of thing for me. :smile: Perhaps @j-friedrich may have some thoughts here he knows this much better than I do.

Edit added: feel free to email me we can set up a discussion, and maybe get Johannes in the loop for this it might be good to talk!

**Reply by bashsyr** [reply]
of course. I'd love to.
I've seen the maxIters parameter somewhere while digging in the parameter dict but I never rally understood how it stops

**Comment by j-friedrich** [comment]
@bashsyr I looked into that a bit [here](http://www.stat.columbia.edu/~johannes/docs/nips2015.pdf). HALS alternates between updating the temporal and spatial components, each involves solving a convex NNLS problem. I think CaImAn does merely 2 outer iterations by default. Each NNLS problem is itself solved iteratively using block-coordiate descent, and 5 inner iterations are usually enough (if there were no overlaps 1 iteration would already be exact). You'd need to check the code (update_temporal and update_spatial), CaImAn may perform less iterations if the components stop changing earlier, e.g. using a condition like $||A_i-A_{i-1}|| / ||A_i|| < tol$ with iteration numer $i$.
As aside, using accelerated projected gradient decent as [FIOLA](https://www.nature.com/articles/s41592-023-01964-2) does, requires more - about 30 - inner iterations, but on a GPU each iteration is faster.

**Reply by bashsyr** [reply]
Thank you so much.
This helps A lot :)
